No multiarch out of tree

Thaodan opened the pull request:
# [Remove multilib for x86_64 and use autotools variables](https://github.com/sailfishos/scratchbox2/pull/32/commits/d34c07703d1a4824749302d3f91d75995c10c0b6)
[d34c077](https://github.com/sailfishos/scratchbox2/pull/32/commits/d34c07703d1a4824749302d3f91d75995c10c0b6)

The builtin multilib support as unused. It's cleaner to built
libsb2 for 32bit targets in a 32bit worker.
The side effect of these changes is that make file pickup
the correct directories passed to configure in %configure.

# Automatically detect out of tree builds
[c73c772](https://github.com/sailfishos/scratchbox2/pull/32/commits/c73c772539f8d63ca089bc757789d8ba9300181d)

In case the root Makefile isn't in PWD assume that the Makefiles
directory is the source directory.

# [Use autotools builtin system to rerun configure instead of our own](https://github.com/sailfishos/scratchbox2/pull/32/commits/04f90b4df583794f4e63bf95183418448a7b4696)
[04f90b4](https://github.com/sailfishos/scratchbox2/pull/32/commits/04f90b4df583794f4e63bf95183418448a7b4696)

By detecting if configure.ac has changed and thus configure will change
in the following step we can run autogen.sh and config.status to rerun
configure with the previously set built options.

The change removes the need for .configure file, correctly detects
if configures sources changed and doesn't run configure again without
all the previously passed options.


Contains changes from #22 and #23.

Thaodan committed:
[sb2] Avoid errors in out of tree builds for files from $(OBJDIR)

Previously in some instance the assumption was $(SRCDIR) = $(OBJDIR).
If $(OBJDIR) isn't equal to $(SRCDIR) these files were not found anymore.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[utils] Use va_args in open_nomap_nolog to match wrappers. JB#61866

Resolves error when using a compiler with stricter settings such as
when using LTO.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[preload] Adapt glibc in open ignore 'mode' with O_TMPFILE in flags. JB#61866

Both open and openat load their last argument 'mode' lazily, using
va_arg() only if O_CREAT is found in oflag. This is wrong, mode is also
necessary if O_TMPFILE is in oflag.

Adapt __OPEN_NEEDS_MODE(oflag) present since glibc 2.22 to solve the
issue.

Read more down below:
https://sourceware.org/git/?p=glibc.git;a=commit;h=65f6f938cd562a614a68e15d0581a34b177ec29d

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Don't run make twice

Once in autogen.sh and once in %make

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[preload] Fix generated dependency order

Previously preload/exported.h depended on preload/wrappers.c even thou
both where generated from interface.master by gen-interface.pl.

Refactor so that all generated preloader files depend on the interface
and the gen-interface script.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[llbuild] Refactor import_dir to match other Makefiles

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Remove multilib for x86_64 and use autotools variables. JB#61868

The builtin multilib support as unused. It's cleaner to built
libsb2 for 32bit targets in a 32bit worker.
The side effect of these changes is that make file pickup
the correct directories passed to configure in %configure.

Also now we can use regular build macros.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Remove multilib for x86_64 and use autotools variables. JB#61868

The builtin multilib support was unused. It's cleaner to built
libsb2 for 32bit targets in a 32bit worker.
The side effect of these changes is that make file pickup
the correct directories passed to configure in %configure.

Also now we can use regular build macros.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Use autotools builtin system to rerun configure instead of our own

By detecting if configure.ac has changed and thus configure will change
in the following step we can run autogen.sh and config.status to rerun
configure with the previously set built options.

The change removes the need for .configure file, correctly detects
if configures sources changed and doesn't run configure again without
all the previously passed options.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Use autotools builtin system to rerun configure instead of our own

By detecting if configure.ac has changed and thus configure will change
in the following step we can run autogen.sh and config.status to rerun
configure with the previously set built options.

The change removes the need for .configure file, correctly detects
if configures sources changed and doesn't run configure again without
all the previously passed options.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Automatically detect out of tree builds

In case the root Makefile isn't in PWD assume that the Makefiles
directory is the source directory.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Automatically detect out of tree builds

In case the root Makefile isn't in PWD assume that the Makefiles
directory is the source directory.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
packaging: Cleanup spec file

Remove obsolete and unused things.
Own all directories correctly (this will be an error in later RPM versions).

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
packaging: Cleanup spec file

Remove obsolete and unused things.
Own all directories correctly (this will be an error in later RPM versions).

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
sb2: Fix errors found shellcheck SC2166, SC2181, SC2048 and SC2012

Remove useless checks like:
```
command
if [ $# -eq 0 ] ; then
    error
fi
```

Or replace obsolescent -a and -o with `[ ] && [ ] ` and `[ ] || [ ]`.

https://github.com/koalaman/shellcheck/wiki/SC2166
https://github.com/koalaman/shellcheck/wiki/SC2181

Thaodan committed:
sb2: Fix errors found shellcheck SC2166, SC2181, SC2048 and SC2012

Remove useless checks like:
```
command
if [ $# -eq 0 ] ; then
    error
fi
```

Or replace obsolescent -a and -o with `[ ] && [ ] ` and `[ ] || [ ]`.

https://github.com/koalaman/shellcheck/wiki/SC2166
https://github.com/koalaman/shellcheck/wiki/SC2181

Thaodan committed:
[sb2] Check if default mode for target exists, error out earlier

Doing so exits with a clearer error message then errors about sb2d
failing to start later such as these:

Runtime-Error loading /tmp/sb2-bidar-20240329-145433.AwoNRb/lua_scripts/init.lua (...dar-20240329-145433.AwoNRb/lua_scripts/argvenvp_misc.lua:28: bad argument #1 to 'files' (/tmp/sb2-bidar-20240329-145433.AwoNRb/wrappers.sdk-build: No such file or directory))
sb2: Error: startup of sb2d failed.

Tested-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Check if default mode for target exists, error out earlier

Doing so exits with a clearer error message then errors about sb2d
failing to start later such as these:

Runtime-Error loading /tmp/sb2-bidar-20240329-145433.AwoNRb/lua_scripts/init.lua (...dar-20240329-145433.AwoNRb/lua_scripts/argvenvp_misc.lua:28: bad argument #1 to 'files' (/tmp/sb2-bidar-20240329-145433.AwoNRb/wrappers.sdk-build: No such file or directory))
sb2: Error: startup of sb2d failed.

Tested-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Take directory variables from configure. JB#61868

To make it easier to detect where Scratchbox2 helper scripts
and binaries are found on the host and partially on the
tooling/target.
This is especially important for helper binaries such as sb2dctl.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Take directory variables from configure. JB#61868

To make it easier to detect where Scratchbox2 helper scripts
and binaries are found on the host and partially on the
tooling/target.
This is especially important for helper binaries such as sb2dctl.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
sb2: Remove trailing slash

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
sb2: Remove trailing slash

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Take version from autoconf to avoid unintended rebuilds

The version used in the Makefile and version header was extracted
using git with a phony target which resulted in unintended rebuilds.

Using the autoconf script avoids these as now the version only updates
when the autoconf script changes.

Further the version originally set in the autoconf was also out of
date. This change fixes both of these issues.
The script in build-aux was taken from Gnulib.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Take version from autoconf to avoid unintended rebuilds

The version used in the Makefile and version header was extracted
using git with a phony target which resulted in unintended rebuilds.

Using the autoconf script avoids these as now the version only updates
when the autoconf script changes.

Further the version originally set in the autoconf was also out of
date. This change fixes both of these issues.
The script in build-aux was taken from Gnulib.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Clean correctly without breaking builds and keeping the tree dirty

Previously config.status was also clean as part of the clean making
rebuilds impossible without calling configure again.
Clean isn't supposed to clean the generated build- or
distribution-system files.

Further somefiles that should be clean where not cleaned at all.
This change also aligns the name of the clean target which removes
the distribution to the one configure mentions when it fails.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Clean correctly without breaking builds and keeping the tree dirty

Previously config.status was also clean as part of the clean making
rebuilds impossible without calling configure again.
Clean isn't supposed to clean the generated build- or
distribution-system files.

Further somefiles that should be clean where not cleaned at all.
This change also aligns the name of the clean target which removes
the distribution to the one configure mentions when it fails.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Make sure config.h is updated and cleaned

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Make sure config.h is updated and cleaned

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Avoid rebuilds when configure.ac changes but not it's output

When configure.ac or config.mak.in change but not their output their
will be repeated rebuilds as the timestamp their output files such as
config.mak stay older than their dependency.

The timestamp file's timestamp will always update after the targets
dependency changed.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
[sb2] Avoid rebuilds when configure.ac changes but not it's output

When configure.ac or config.mak.in change but not their output their
will be repeated rebuilds as the timestamp their output files such as
config.mak stay older than their dependency.

The timestamp file's timestamp will always update after the targets
dependency changed.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

martyone commented on utils/sb2-session:
```diff
@@ -142,21 +142,23 @@ attach_session()
-		printf "%s\t%s\t%s\n" "$session_file" "$session_target" \
-		    "$session_descr"
+	for s in $session_save_dir/session.* ; do
+		if [ -e $s ] ; then
```
I would prefer doing `[ -e $s ] || continue` instead.

martyone commented on utils/sb2:
```diff
@@ -356,17 +354,17 @@ clone_target_root_dir_from()
-	if [ $? != 0 ]; then
+
+	if ! (cd $source_directory; find . -depth -print |
+			  cpio -pamd $SBOX_SESSION_DIR/target_root) ; then
```
While touching these it would be nice to fix it to have operators (`|`, `||`, `&&` etc.) at the beginnings of continuation lines.

```suggestion
	if ! (cd $source_directory; find . -depth -print \
			| cpio -pamd $SBOX_SESSION_DIR/target_root) ; then
```

martyone commented on utils/sb2-session:
```diff
@@ -225,11 +227,11 @@ case "$command" in
 		if [ $inside_sb2 -ne 0 ]; then
-			echo -n "'attach' command cannot be run from inside "
+			printf "%s" "'attach' command cannot be run from inside "
 			echo "sb2 session"
```
I'd suggest to pass the message directly and replace also the second `echo`. More occurrences elsewhere.

martyone commented on utils/sb2-session:
```diff
@@ -191,9 +193,9 @@ set_env_var()
 show_env_vars()
 {
 	if [ $inside_sb2 -eq 1 ]; then
-		count=$( ls $SBOX_SESSION_DIR/env_vars/* 2>/dev/null | wc -l )
+		count=$(find "$SBOX_SESSION_DIR/env_vars" -print x | wc -c )
 		if [ "$count" -gt 0 ]; then
-			cat $SBOX_SESSION_DIR/env_vars/*
+			cat "$SBOX_SESSION_DIR"/env_vars/*
 		fi
```
I think this does not work as expected. First, `-printf` would have to be used or `x` is treated as a path. Second, it also prints the `env_vars` dir itself. Also maybe replace the whole thing with

```
		find "$SBOX_SESSION_DIR/env_vars" -type f -exec cat '{}' ';'
```

Thaodan replied:
Why cat? That would open each file. I will change to printf and match only files.

martyone replied:
I think you misread something. Or if your concerns are about forking `cat` for each file, then of course, we can optimize with `find ...|xargs cat` :)

Thaodan replied:
My concerns are about opening each file with cat when the point of this command is only to  count amount of files.

martyone replied:
Check it again :) If the count is greater than zero, then all files are `cat` to the stdout. Hence the command I suggested, which does it all in one "step".

Thaodan replied:
Ah I get it now. Thanks.

martyone commented on utils/sb2-upgrade-config:
```diff
@@ -99,8 +99,7 @@ update_toolchain_configs()
 			-- \
-			$compiler_path
-		if [ $? != 0 ]; then
+			$compiler_path; then
```
I am not sure about these. Unless indented once more, readability is worse then with the original.

Thaodan replied:
Because of the deeper nesting?

martyone commented on utils/sb2:
```diff
@@ -233,6 +233,10 @@ load_configuration()
 	fi
 
+	if [ ! -d "$SBOX_DIR/share/scratchbox2/modes/$SBOX_MAPMODE" ]; then
+		exit_error "Configure mode doesn't '$SBOX_MAPMODE' for target doesn't exist, aborting."
```
Typo?

martyone commented on utils/sb2-config-gcc-toolchain.sh.in:
```diff
@@ -39,7 +31,7 @@ Options:
     -h                print this help
     -m mapping_mode   target uses mapping_mode as default mode
-    -S SBOX_DIR       define value for SBOX_DIR
+    -S SBOX_DATADIR       define value for SBOX_DIR
```
Bad spacing?

martyone commented on utils/sb2-config-gcc-toolchain.sh.in:
```diff
@@ -169,8 +161,8 @@ esac
-if [ -z "$SBOX_DIR" ]; then
-	SBOX_DIR=$(readlink -f $(dirname $(readlink -f $my_path))/..)
+if [ -z "$SBOX_DATADIR" ]; then
+    SBOX_DATADIR=@datadir@
```
Bad indent?

martyone commented on utils/sb2-init.sh.in:
```diff
@@ -231,10 +223,12 @@ configure_toolchains()
 
-if [ -z "$SBOX_DIR" ]; then
-	SBOX_DIR=$(readlink -f $(dirname $(readlink -f $my_path))/..)
+if [ -z "$SBOXDATA_DIR" ]; then
```
typo? SBOX_DATADIR. Similar below.

martyone commented on utils/sb2.sh.in:
```diff
@@ -918,13 +908,18 @@ locate_shell()
+            fi
+        done
+	fi
+	printf "$ld_so_conf"
```
Using spaces for indentation by mistake? More occurrences elsewhere.

martyone commented on utils/sb2.sh.in:
```diff
@@ -956,18 +951,22 @@ write_libsb2_and_ld_so_state_to_exec_config()
+	        fi
+            break
+        fi
+    done
```
Bad indent. Consider using `[ -d $rootdir$liblocation ] || continue` to avoid extra nesting.

martyone commented on utils/sb2.sh.in:
```diff
@@ -1153,7 +1152,12 @@ END
+        if [ -d $liblocation ] ; then
+            break
+        fi
+    done
```
Bad indent. More occurrences elsewhere.

martyone commented on utils/sb2.sh.in:
```diff
@@ -1511,20 +1515,40 @@ link_wrappers_for_mapmode()
-					$SBOX_SESSION_DIR/wrappers.$w_mode 2>/dev/null
+			if [ "$SBOX_WRAPPERS" = 'ALL' ]; then
+                for wrapper in $SBOX_DATADIR/scratchbox2/wrappers/* ; do
+                    if [ -e $wrapper ]; then
```
I'd prefer `foo || continue`. More occurrences below.

martyone commented on configure.ac:
```diff
@@ -3,7 +3,12 @@
-AC_INIT([sb2],[1.99b],[lle@rahina.org])
+AC_INIT([SB2],
+        m4_esyscmd([build-aux/git-version-gen --prefix '' .tarball-version]),
+        [bjorn.bidar@jolla.com],
```
https://forum.sailfishos.org/ could be more appropriate for bug reporting or maybe leave empty?

Thaodan replied:
I will check.

Thaodan replied:
Changed to  a link to Sailfish Docs.

martyone commented on .gitignore:
```diff
@@ -7,6 +7,8 @@
 include/config.h
 include/config.h.in
+stamp-h
+stamp-h.in
```
Are the `.in` files actually used?

Thaodan replied:
Yes  they are later down the line by configure.

martyone replied:
Ah now I see how.

martyone commented on Makefile:
```diff
@@ -75,15 +75,23 @@ do-all: $(targets)
-$(DIST_FILES): $(SRCDIR)/configure $(SRCDIR)/config.mak.in
+# Use a proxy timestamp file here since config.mak and config.h
+# don't update unless their contents change when the dependencies
+# that generate them change, avoid repeated rebuilds after changing
+# files such as configure.ac e.g. after checking out a new branch
```
Are you sure this is not going to be a kind of gotcha after all? Is it still easier and more transparent than the original approach before "Use autotools builtin system to rerun configure instead of our own"? (I haven't checked thoroughly.)

Thaodan replied:
The old system did not take care of rebuilding autoconf at all besides creating if it didn't exist.

Using config.status is the intended way of regenerating the build system. The issue lies with the timestamps not updating when the result of configure.ac or config.h.in did not change but the timestamp of the source flie did (e.g. configure.ac did).

Using the timestamp files ensures that configure is regenerated if configure.ac changed once and not every time make is invoked since configure.ac  will be newer if e.g. git changes it's timestamp or other changes to the file such as comments where added which do not change it's output file.

The same applies to config.h of course.

For more context read:
https://www.gnu.org/software/emacs/manual/html_node/autoconf/Automatic-Remaking.html

martyone replied:
OK, consider adding "See Automatic Remaking in autoconf's manual" to the comment. The link did not open.

Thaodan replied:
Oh I see it that now. I seems that info gave me the wrong link to the HTML version of the manual.

Here is the right link:
https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.72/html_node/Automatic-Remaking.html

martyone replied:
Yup, I was able to quickly find it. I just meant it would be good to refer it in the comment and maybe use natural language instead of the version-specific URL.

direc85 committed:
sb2: Move logic to functions, session checks to parser, shorten code

Co-authored-by: Matti Viljanen <matti.viljanen@jolla.com>
Co-authored-by: Björn Bidar <bjorn.bidar@jolla.com>
Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
sb2: Move logic to functions, session checks to parser, shorten code

Co-authored-by: Matti Viljanen <matti.viljanen@jolla.com>
Co-authored-by: Björn Bidar <bjorn.bidar@jolla.com>
Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

direc85 commented on Makefile:
```diff
@@ -169,19 +169,19 @@ install-noarch: all
+	$(Q)install -m 755 $(OBJDIR)/utils/sb2-upgrade-config $(DESTDIR)$(datadir)/scratchbox2/scripts/sb2-upgrade-config
 	$(Q)install -m 755 $(SRCDIR)/utils/sb2-parse-sb2-init-args $(DESTDIR)$(datadir)/scratchbox2/scripts/sb2-parse-sb2-init-args
-	$(Q)install -m 755 $(SRCDIR)/utils/sb2-config-gcc-toolchain $(DESTDIR)$(datadir)/scratchbox2/scripts/sb2-config-gcc-toolchain
+	$(Q)install -m 755 $(OBJDIR)/utils/sb2-config-gcc-toolchain $(DESTDIR)$(datadir)/scratchbox2/scripts/sb2-config-gcc-toolchain
```
I would group the install lines in these blocks by `$(OBJDIR)` and `$(SRCDIR)` to highlight the difference.

Thaodan replied:
Currently they are grouped by subfolders. What do you think is better?

direc85 replied:
Either works, this is a nitpick anyway :)

direc85 commented on utils/sb2-session:
```diff
@@ -216,30 +208,23 @@ else
 fi
 
+case $command in
+    attach|save|delete|setenv|showenv)
```
`save` command can only be run from inside sb2 session.

mlehtima commented on rpm/scratchbox2.spec:
```diff
@@ -1,11 +1,10 @@
 Release:	0
 Source:		%{name}-%{version}.tar.gz
-Prefix:		/usr
 ExclusiveArch:	%{ix86} %{x86_64}
```
Fix this while at it, x86_64 macro is only available in rpm 4.19 or newer. So replace %{x86_64} with x86_64 for now.

Thaodan committed:
packaging: Also include x86_64 in ExclusiveArch

In case not available.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
packaging: Also include x86_64 in ExclusiveArch

In case not available.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

direc85 commented on rpm/scratchbox2.spec:
```diff
@@ -1,12 +1,11 @@
 Source:		%{name}-%{version}.tar.gz
-Prefix:		/usr
-ExclusiveArch:	%{ix86} %{x86_64}
+ExclusiveArch:	%{ix86} %{x86_64} x86_64
```
Both %{x86_64} and x86_64 here now?

Thaodan replied:
Either %{x86_64} is expanded or empty. Either way x86_64 works as a fallback.

direc85 commented on utils/sb2-session:
```diff
@@ -142,21 +142,23 @@ attach_session()
 		session_file=$(basename $s)
 		session_descr=$( grep SBOX_SESSION_DESCRIPTION $s | \
-		    sed 's/SBOX_SESSION_DESCRIPTION="\([^"]*\)"/\1/' )
+							 sed 's/SBOX_SESSION_DESCRIPTION="\([^"]*\)"/\1/' )
```
Weird intent? Spaces vs. tabs I think?

Thaodan committed:
Add EditorConfig and dir-locals

Sets basic indentation across editors and reviews on Github.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

Thaodan committed:
Add EditorConfig and dir-locals

Sets basic indentation across editors and reviews on Github.

Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

direc85 commented on Makefile:
```diff
@@ -77,50 +60,86 @@ CXXFLAGS =
+# or follow:
+# https://www.gnu.org/prep/standards/html_node/Makefile-Conventions.html
+
+all: $(DIST_FILES) $(OBJDIR)/include/config.h .WAIT do-all
```
Since `.WAIT` [requires make 4.4](https://lwn.net/Articles/913253/), would this work instead?
```suggestion
all: $(DIST_FILES) $(OBJDIR)/include/config.h do-all
.NOTPARALLEL: all
```

direc85 replied:
Or make could be told that there's a dependency, something like this?

```suggestion
all: do-first do-all
do-first: $(DIST_FILES) $(OBJDIR)/include/config.h
do-all: do-first $(targets)
```

Thaodan committed:
preload: Remove mentions of ldexportlist or replace with export.map

ldexportlist was removed earlier for export.map. Replace with
export.map where not done already. Remove obsolete references such as
the target in preload/Makefile.

Fixes: 20180ac890cd54ccc1ee1b450a660681e87ce1bd
Signed-off-by: Björn Bidar <bjorn.bidar@jolla.com>

direc85 commented:
6c5679af49a0da9959e0620468381bbd0879855b doesn't compile on Patform SDK:
```text
/srv/mer/toolings/SailfishOS-devel/opt/cross/bin/i486-meego-linux-gnu-ld:/home/mattiv/PSDK/scratchbox2/build/preload/export.map: file format not recognized; treating as linker script
/srv/mer/toolings/SailfishOS-devel/opt/cross/bin/i486-meego-linux-gnu-ld:/home/mattiv/PSDK/scratchbox2/build/preload/export.map:1: syntax error
collect2: error: ld returned 1 exit status
```

Thaodan commented:
> [6c5679a](https://github.com/sailfishos/scratchbox2/commit/6c5679af49a0da9959e0620468381bbd0879855b) doesn't compile on Patform SDK:
> 
> ```
> /srv/mer/toolings/SailfishOS-devel/opt/cross/bin/i486-meego-linux-gnu-ld:/home/mattiv/PSDK/scratchbox2/build/preload/export.map: file format not recognized; treating as linker script
> /srv/mer/toolings/SailfishOS-devel/opt/cross/bin/i486-meego-linux-gnu-ld:/home/mattiv/PSDK/scratchbox2/build/preload/export.map:1: syntax error
> collect2: error: ld returned 1 exit status
> ```

That's fixed now.

martyone commented on utils/sb2:
```diff
@@ -272,9 +272,7 @@ load_configuration()
-			$SBOX_DIR/share/scratchbox2/scripts/sb2-upgrade-config \
-				$SBOX_TARGET
-			if [ $? != 0 ]; then
+            if ! $SBOX_DIR/share/scratchbox2/scripts/sb2-upgrade-config $SBOX_TARGET; then
```
Bad indent

martyone commented on utils/sb2:
```diff
@@ -1396,16 +1394,14 @@ initialize_new_sb2_session()
 		date_and_time_now=$(date +%Y%m%d-%H%M%S)
-		SBOX_SESSION_DIR=$(mktemp -d /tmp/sb2-$USER-$date_and_time_now.XXXXXX)
-		if [ $? != 0 ]; then
+        if ! SBOX_SESSION_DIR=$(mktemp -d /tmp/sb2-$USER-$date_and_time_now.XXXXXX) ; then
```
Bad indent

martyone commented on utils/sb2-config:
```diff
@@ -167,16 +166,16 @@ function unsetenv_var()
 {
 	if [ -d $SBOX_CONFIG_DIR/env_vars ]; then
-		count=$( ls $SBOX_CONFIG_DIR/env_vars/* 2>/dev/null | wc -l )
+	    count=$(find "$SBOX_SESSION_DIR/env_vars" -print x | wc -c )
```
Bad indent

martyone commented on utils/sb2-exitreport:
```diff
@@ -14,13 +14,13 @@
-if [ -z "$SBOX_MAPPING_LOGFILE" -o \
-     -z "$exit_reason" -o -z "$exit_status" ]; then
+if [ -z "$SBOX_MAPPING_LOGFILE" ] || \
+	   [ -z "$exit_reason" ] || [ -z "$exit_status" ]; then
```
Bad indent

martyone commented on utils/sb2.sh.in:
```diff
@@ -1497,40 +1500,56 @@ join_existing_session()
+	    mkdir $SBOX_SESSION_DIR/wrappers.$w_mode
+
+	    if [ "$SBOX_WRAPPERS" = 'ALL' ]; then
+		for wrapper in $SBOX_DATADIR/ ; do
```
Feels like something is missing here.

martyone commented on utils/sb2.sh.in:
```diff
@@ -1497,40 +1500,56 @@ join_existing_session()
+		    wrapper_found=nil
+		    for wrapper_dir in $SBOX_DATADIR/scratchbox2 \
+					   $sbox_liblocations \
+				       ; do
```
Typo?

martyone commented on utils/sb2.sh.in:
```diff
@@ -1497,40 +1500,56 @@ join_existing_session()
+			wrapper_found=wrapper_found=t
+			break
+		    done
+		    if [ $wrapper_found = nil ] ; then
```
I would prefer simply using empty value as False and 1 as True.
```
wrapper_found=
if ...
    wrapped_found=1
fi
if [ ! $wrapper_found ]; then
...
```

martyone commented on utils/sb2.sh.in:
```diff
@@ -1497,40 +1500,56 @@ join_existing_session()
+			[ -f $wrapper_dir/wrappers/$ammw ] || continue
+			ln -s $wrapper_dir/wrappers/$ammw \
+			   $SBOX_SESSION_DIR/wrappers.$w_mode
+			wrapper_found=wrapper_found=t
```
Typo?

martyone commented on rpm/scratchbox2.spec:
```diff
@@ -37,6 +37,11 @@ Scratchbox2 man pages.
 %prep
 %autosetup
+# Tell autoconf the package version
+# Note we don't strip the version here not to not remove the indicator
```
Extra not

martyone commented on Makefile:
```diff
@@ -111,15 +115,25 @@ do-all: $(targets)
+# don't update unless their contents change when the dependencies
+# that generate them change, avoid repeated rebuilds after changing
+# files such as configure.ac e.g. after checking out a new branch.
+# Read "(autoconf) Automatic Remaking" in Info or follow:x
```
Typo?

martyone commented on utils/sb2-session:
```diff
@@ -216,30 +209,25 @@ else
-		. $SBOX_SESSION_DIR/sb2-session.conf
-
+		[ $inside_sb2 -eq 1 ] \
+			|| die die "'save' command can only be run from inside sb2 session"
```
Afraid it could return? :)

Thaodan replied:
Just to be sure it's dead.

martyone reviewed (commented):
Noticed just some stylish issues.

In addition to those, it would really be nice, now that we are touching all the multi-line expressions, to ensure operators are at the beginnings of the continuation lines and not hanging at the ends.

I am not 100% sure readability improves by eliminating the indirect checks of exit codes `[ $? -eq 0 ]`. At least I would ensure that the continuation lines of a multi-line `if` conditions use two levels of indentation, so that it is immediately clear where the body starts. With that, we do not need to break the convention if placing the `then` keyword at the end of the line.

```
if my super long \
        command as a condition \
        && another one \
            |further filtered \
        || alternatively this one; then
    the body starts here
    and continues
fi
```

pvuorela commented on rpm/scratchbox2.spec:
```diff
@@ -1,12 +1,11 @@
-Summary:	Scratchbox2 crosscompiling environment
+Summary:	Crosscompiling environment
 License:	LGPLv2
 URL:		https://git.sailfishos.org/mer-core/scratchbox2
```
Hey btw this could be fixed in whatever PR that's getting first in. One rare remaining case of mer-core.

Thaodan committed:
[preloader] Don't link lua into libsb2

When Lua was unbundled, i.e. removed from libsb2, liblua was
accidentally linked to lisb2.
That added liblua as a dependency to targets even thou it wasn't used in them.

Fixes: c967c03c5be743eb23587f00f0dfe9a6c06ece7a

Thaodan committed:
[preloader] Don't link lua into libsb2

When Lua was unbundled, i.e. removed from libsb2, liblua was
accidentally linked to lisb2.
That added liblua as a dependency to targets even thou it wasn't used in them.

Fixes: c967c03c5be743eb23587f00f0dfe9a6c06ece7a

Thaodan committed:
[sb2] Redo version handling, determine soname at runtime

Before the version was defined in an extra header but also defined in
config.h which includes the version anyway as it's put in there through
the autoconf script.
Doing so avoid race conditions during parallel builds.

The version header also defined the sover. The sover was defined
in multiple places before, both in the build as well as in the shell.
This inconsistency meant that in some places it said the sover was one
in others it was two.

Now the sover is set in configure.ac once for both C code aswell as the
rest.

The actual soname is set once during build for itself and target
initialization.
After that it's determined at runtime through the ELF interpreter.

Thaodan committed:
[sb2] Redo version handling, determine soname at runtime

Before the version was defined in an extra header but also defined in
config.h which includes the version anyway as it's put in there through
the autoconf script.
Doing so avoid race conditions during parallel builds.

The version header also defined the sover. The sover was defined
in multiple places before, both in the build as well as in the shell.
This inconsistency meant that in some places it said the sover was one
in others it was two.

Now the sover is set in configure.ac once for both C code aswell as the
rest.

The actual soname is set once during build for itself and target
initialization.
After that it's determined at runtime through the ELF interpreter.

Thaodan committed:
[sb2] Define the preloader per policy context

Enable the preloader to be different depending on the policy context,
i.e. for backwards compatibility and for testing.

Thaodan committed:
[sb2] Define the preloader per policy context

Enable the preloader to be different depending on the policy context,
i.e. for backwards compatibility and for testing.
